packages/otp2-tile-overlay/src/index.tsx

```diff
@@ -0,0 +1,178 @@
+import { StopPopup } from "@opentripplanner/stops-overlay"
+import React, { useEffect, useState } from "react"
+// eslint-disable-next-line prettier/prettier
+import type { EventData } from "mapbox-gl"
+import { LAYER_PAINT } from "./util"
+
+const OTP2TileLayerWithPopup = ({
+  configCompanies,
+  id,
+  network,
+  setLocation,
+  setViewedStop,
+  type
+}: {
+  /**
+   * Optional configuration item which allows for customizing properties of scooter and
+   * bikeshare companies
+   */
+  configCompanies?: ConfiguredCompany[]
+  id: string
+  name?: string
+  /**
+   * If `network` is specified, the layer will be filtered to only show vehicles from
+   * that network
+   */
+  network?: string
+  /**
+   * A method fired when a stop is selected as from or to in the popup
+   */
+  setLocation: (location: MapLocationActionArg) => void
+  /**
+   * A method fired when the stop viewer is opened in the popup
+   */
+  setViewedStop: ({ stopId }: { stopId: string }) => void
+  type: string
+}): JSX.Element => {
+  const { current: map } = useMap()
+
+  // TODO: handle this complex type: it can be a stop, a station, and some extra fields too
+  const [clickedEntity, setClickedEntity] = useState<any>(null)
+  useEffect(() => {
+    map?.on("mouseenter", id, () => {
+      map.getCanvas().style.cursor = "pointer"
+    })
+    map?.on("mouseleave", id, () => {
+      map.getCanvas().style.cursor = ""
+    })
+    map?.on("click", id, (event: EventData) => {
+      const { sourceLayer } = event.features?.[0]
+      const synthesizedEntity = {
+        ...event.features?.[0].properties,
+        lat: event.lngLat.lat,
+        lon: event.lngLat.lng,
+        sourceLayer
+      }
+
+      if (sourceLayer === "stops" || sourceLayer === "stations") {
+        setClickedEntity(synthesizedEntity)
+      }
+      if (
+        sourceLayer === "rentalVehicles" ||
+        sourceLayer === "rentalStations"
+      ) {
+        setClickedEntity({
+          // GraphQL field not in the tile info, but we can deduce it
+          isFloatingBike:
+            sourceLayer === "rentalVehicles" &&
+            synthesizedEntity.formFactor === "BICYCLE",
+          // GraphQL field not in the tile info, but we can deduce it
+          isFloatingVehicle:
+            sourceLayer === "rentalVehicles" &&
+            synthesizedEntity.formFactor === "SCOOTER",
+          // OTP1 compatibility -- will get overwritten if possible
+          networks: [synthesizedEntity.network],
+          ...synthesizedEntity,
+          // OTP1 compatibility
+          bikesAvailable: synthesizedEntity.vehiclesAvailable,
+          // OTP1 compatibility
+          x: synthesizedEntity.lon,
+          // OTP1 compatibility
+          y: synthesizedEntity.lat
+        })
+      }
+    })
+  }, [id, map])
+
+  return (
+    <>
+      <Layer
+        filter={network ? ["all", ["==", "network", network]] : ["all"]}
+        id={id}
+        key={id}
```

Comment:
Do we need a `key` prop?

Reply:
Yes. `base-map` removes the fragment

Reply:
See my comment on the main comment about this. This annoying complexity has to live somewhere -- the only way to avoid using this method is to re-write the base map layer selector to require additional configuration which I am very opposed to.